### PR TITLE
EQL: [Tests] Add correctness integration tests

### PR DESCRIPTION
**Pull request by mark-vieira**
Add a new gradle module under `eql/qa` which runs and validates a set of
queries  over a 4m event dataset (restored from a snapshot residing in a
gcs bucket). The results are providing by running the exact set of queries
with Python EQL against the same dataset.

### CODE REVIEW

File: x-pack/plugin/eql/qa/correctness/src/javaRestTest/java/org/elasticsearch/xpack/eql/EsEQLCorrectnessIT.java [hunk 1]
```diff
@@ -0,0 +1,187 @@
...
+
+    private static final String PARAM_FORMATTING = "%1$s";
+    private static final String QUERIES_FILENAME = "queries.toml";
+    private static final String INDEX_NAME = "mitre";
+    private static final int FETCH_SIZE = 10000;
+    private static final String GCS_REPO_NAME = "eql_correctness_gcs_repo";
+    private static final String SNAPSHOT_NAME = "mitre-snapshot";
+    private static final String GCS_BUCKET_NAME = "matriv-gcs";
+    private static final String GCS_BASE_PATH = "mitre-data";
+    private static final String GCS_CLIENT_NAME = "eql_test";
```
Comment (costin) [inline x-pack/plugin/eql/qa/correctness/src/javaRestTest/java/org/elasticsearch/xpack/eql/EsEQLCorrectnessIT.java:51]: These can be externalized into a properties file.
The "matriv-gcs" should be changed to something more generic.

File: x-pack/plugin/eql/qa/correctness/src/javaRestTest/java/org/elasticsearch/xpack/eql/EsEQLCorrectnessIT.java [hunk 2]
```diff
@@ -0,0 +1,187 @@
...
+        /* For debugging
+        for (int i = 0; i < spec.filters().length; i++) {
+            String filterQuery = spec.filters()[i];
+            EqlSearchRequest eqlSearchRequest = new EqlSearchRequest(INDEX_NAME, filterQuery);
+            eqlSearchRequest.eventCategoryField("event_type");
+            eqlSearchRequest.size(100000);
+            EqlSearchResponse response = client.eql().search(eqlSearchRequest, commonRequestOptions);
+            assertEquals("Failed to match filter counts for query No: " + queryNo + " filterCount: " + i,
+                    spec.filterCounts()[i], response.hits().events().size());
+        } */
```
Comment (costin) [inline x-pack/plugin/eql/qa/correctness/src/javaRestTest/java/org/elasticsearch/xpack/eql/EsEQLCorrectnessIT.java:161]: This should be moved into a separate method and be guarded by an if statement for a bool flag which can be turned on/off if needed.

File: x-pack/plugin/eql/qa/correctness/src/javaRestTest/java/org/elasticsearch/xpack/eql/EsEQLCorrectnessIT.java [hunk 3]
```diff
@@ -0,0 +1,187 @@
...
+
+        EqlSearchRequest eqlSearchRequest = new EqlSearchRequest(INDEX_NAME, spec.query());
+        eqlSearchRequest.eventCategoryField("event_type");
+        eqlSearchRequest.tiebreakerField("serial_id");
+        eqlSearchRequest.size(2000);
+        eqlSearchRequest.fetchSize(FETCH_SIZE);
```
Comment (costin) [inline x-pack/plugin/eql/qa/correctness/src/javaRestTest/java/org/elasticsearch/xpack/eql/EsEQLCorrectnessIT.java:167]: These should be extracted as well to constants to make it easy to tweak.

**Review by costin — APPROVED** [review]
Left some style comments but otherwise LGTM

File: x-pack/plugin/eql/qa/correctness/src/javaRestTest/java/org/elasticsearch/xpack/eql/EsEQLCorrectnessIT.java (hunk 1 above)
Comment (matriv) [reply x-pack/plugin/eql/qa/correctness/src/javaRestTest/java/org/elasticsearch/xpack/eql/EsEQLCorrectnessIT.java:51]: I'd be meaning to do it, but let's get this in as is, and I will rename the bucket soon.

**Comment by costin** [comment]
@matriv LGTM

**Review by astefan — APPROVED** [review]
LGTM code wise, but I would like to see some more documentation behind this test module. Maybe a class-level javadoc on `EsEQLCorrectnessIT` explaining the "debug" part a bit more and where/how the .toml file comes from. The PR description talks about Python tests comparison, but the gradle code behind it doesn't do anything python related. This bit and the lack of details in the code created a bit of confusion for me. Thanks.

**Comment by mark-vieira** [comment]
> LGTM code wise, but I would like to see some more documentation behind this test module.

Perhaps a README in the project folder. I like this as they render in GitHub, are immediately obvious when browsing the source and are just a nicer form factor for non-code documentation. We could explain at a high level how this stuff works, where to access the creds necessary to execute the tests, etc.

**Pull request merged**